src/data/polynomial/algebra_map.lean

```diff
@@ -247,6 +247,10 @@ lemma coeff_zero_eq_aeval_zero' (p : polynomial R) :
   algebra_map R A (p.coeff 0) = aeval (0 : A) p :=
 by simp [aeval_def]
 
+theorem aeval_comm (a : A) (p q : polynomial R) :
+  commute (aeval a p) (aeval a q) :=
+by { refine (_ : _ * _ = _ * _), simp only [←aeval_mul, mul_comm] }
```

Comment:
Isn't this a general statement about `alg_hom`s?

Reply:
Or even just `mul_hom`s: this is just `commute.map mul_comm (aeval a)` I think.

Reply:
@Vierkantor, how hard is changing `commute.map` to take `mul_hom_class`?

Reply:
It should be painless since there's no dot notation on the morphisms. Just replace any `(f : monoid_hom M N)` with `{F : Type*} [monoid_hom_class F M N] (f : F)`. I encourage you to try doing it, to see if the hom classes are easy enough to work with (and also since I'm currently compiling mathlib).

Reply:
@eric-wieser, yes, sure it holds for `alg_hom`s or even `mul_hom`s. How do you want me to proceed? Do you want me to wait for you to change `commute.map`?

Reply:
This works for me:
```
theorem commute.map {M N F : Type*} [has_mul M] [has_mul N] [mul_hom_class F M N] {x y : M}
  (h : commute x y) (f : F) : commute (f x) (f y) :=
by { refine (_ : _ * _ = _ * _), simp only [←map_mul, h.eq] }
```
@Vierkantor Should I just include it in this PR in place of the existing `commute.map`? Note: when I say "it works," I mean it works in place of what I was using, but I haven't built all of mathlib.

Reply:
oooh, actually, it doesn't, but I'm not sure why. That is, the replacement for `commute.map` compiles, but it doesn't work as a stand-in for my application. (The pretty printer says it failed to unify two types that look the same and `pp.all` is too verbose for me to parse.)

Reply:
Let me see if I can diagnose the error...

Reply:
The error seems to be that it can't unify the `commute` lemma's `mul_one_class → has_mul` with `A`'s `semiring → non_unital_semiring → has_mul` because there are slightly too many holes in the inferred type. But we don't need `mul_one_class` at all, only `has_mul`, so this is essentially a hint that we can generalize the lemmas to `mul_hom_class` for free!

Reply:
Hang on, `alg_hom` doesn't yet have a `mul_hom_class` instance, so no wonder there was a type issue! Interesting though that the error was so different...

Reply:
Done. Check out commit 58bbbae097 for the working code, and I'll prepare two PRs for the dependencies.

Reply:
Created #10853 and #10584. Thank you for trying out my new classes, sorry that they didn't work quite well yet!

Reply:
Thanks for the help!  I understand it will take some time to transition into these new classes.